components/config/opts.rs

```diff
@@ -657,11 +657,10 @@ pub fn from_cmdline_args(args: &[String]) -> ArgumentParsingResult {
     }
 
     let cwd = env::current_dir().unwrap();
-    let homepage_pref = PREFS.get("shell.homepage");
     let url_opt = if !opt_match.free.is_empty() {
         Some(&opt_match.free[0][..])
     } else {
-        homepage_pref.as_string()
+        None
```

Comment:
Sounds about right.

The `match` block should be replaced with something that would achieve this:

if `url_opt` is `None`, keep it `None`.
If `url_opt` is `Some(foo)`, `url_opt` should become `Some(parse_url_or_filename(foo).unwrap())`, but if the result of `parse_url_or_filename(foo)` is `Err()`, we should call `args_fail` (which will exit).

Achieving this should be easy with some combinaison of unwrap and map calls (methods of Option and Result types).

`parse_url_or_filename` return a `Result` that is more or less like an `Option`.

Make sure to understand:

* https://doc.rust-lang.org/std/option/
* https://doc.rust-lang.org/std/result/

This should be one or two lines of code. I let you try to understand what this should look like. Manipulating options is a core concept of Rust. It's worth the time understanding it.

Do not hesitate to ask more questions.